File: src/sine_wave_pkg/sine_wave_pkg/sine_wave_publisher.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# Copyright (c) 2023 Alaa
# This code is licensed under the MIT License.
# See the LICENSE file in the project root for license terms.

import rclpy
from rclpy.node import Node
from std_msgs.msg import String, Float64
from rcl_interfaces.msg import SetParametersResult
import math
from sine_wave_pkg.sine_wave_parameters import sine_wave_params


class SineWavePublisher(Node):

    def __init__(self):
        super().__init__('sine_wave_publisher')

        self.param_listener = sine_wave_params.ParamListener(self)
        self.params = self.param_listener.get_params()

        self.amplitude = self.params.amplitude
        self.phase = self.params.phase
        self.angular_frequency = self.params.angular_frequency
        self.frequency = self.params.frequency
        self.sine_wave_topic = self.params.topic_name

        self.add_on_set_parameters_callback(self.parameter_callback)

        self.sine_wave_publisher = self.create_publisher(
            Float64, self.sine_wave_topic, 10)
        self.sine_wave_timer = self.create_timer(
            1.0 / self.frequency, self.sine_wave_publish_message)
        self.angle = 0.0

        self.get_logger().info('SineWavePublisher node has been started.')

    def update_timer(self, new_frequency):
        self.frequency = new_frequency
        self.sine_wave_timer.cancel()
        self.sine_wave_timer = self.create_timer(
            1.0 / self.frequency, self.sine_wave_publish_message)

    def sine_wave_publish_message(self):

        if self.param_listener.is_old(self.params):
            self.params = self.param_listener.get_params()

        msg = Float64()
        msg.data = self.amplitude * \
            math.sin(self.angular_frequency * self.angle + self.phase)
        self.sine_wave_publisher.publish(msg)
        self.angle += 0.0174533

        self.update_timer(self.frequency)

    def parameter_callback(self, params):
        for param in params:
            if param.name == 'amplitude':
                self.amplitude = param.value
            elif param.name == 'phase':
                self.phase = param.value
            elif param.name == 'angular_frequency':
                self.angular_frequency = param.value
            elif param.name == 'frequency':
                self.frequency = param.value
            self.get_logger().info('Parameter %s has been set to: %s' %
                                   (param.name, param.value))
        return SetParametersResult(successful=True)


def main(args=None):
    rclpy.init(args=args)
    node = SineWavePublisher()

    rclpy.spin(node)

    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/sine_wave_pkg/src/sine_wave_publisher.cpp
#include <chrono>
#include <cmath>
#include <memory>
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/float64.hpp>
#include <sine_wave_pkg/sine_wave_parameters.hpp>
#include "rcl_interfaces/msg/set_parameters_result.hpp"

using namespace std::chrono_literals;

class SineWavePublisher : public rclcpp::Node
{
public:
    SineWavePublisher()
        : Node("sine_wave_publisher"), angle_(0.0)
    {
        param_listener_ = std::make_shared<sine_wave_params::ParamListener>(get_node_parameters_interface());
        params_ = param_listener_->get_params();

        publisher_ = this->create_publisher<std_msgs::msg::Float64>(params_.topic_name, 10);
        update_timer();

        // Register the parameter callback
        param_callback_handle_ = this->add_on_set_parameters_callback(
            std::bind(&SineWavePublisher::parameter_callback, this, std::placeholders::_1));
    }

private:
    void publish_sine_wave()
    {
        auto message = std_msgs::msg::Float64();
        message.data = params_.amplitude * std::sin(angle_ * params_.angular_frequency + params_.phase);

        RCLCPP_INFO(this->get_logger(), "Publishing: '%f'", message.data);
        publisher_->publish(message);
        angle_ += 0.1;
    }

    rcl_interfaces::msg::SetParametersResult parameter_callback(const std::vector<rclcpp::Parameter> &params)
    {
        rcl_interfaces::msg::SetParametersResult result;
        result.successful = true;

        for (const auto &param : params)
        {
            if (param.get_name() == "amplitude")
            {
                params_.amplitude = param.as_double();
            }
            else if (param.get_name() == "phase")
            {
                params_.phase = param.as_double();
            }
            else if (param.get_name() == "angular_frequency")
            {
                params_.angular_frequency = param.as_double();
            }
            else if (param.get_name() == "frequency")
            {
                params_.frequency = param.as_double();
                update_timer(); // Update the timer when frequency changes
            }
            RCLCPP_INFO(this->get_logger(), "Parameter %s has been set to: %f", param.get_name().c_str(), param.as_double());
        }

        return result;
    }

    void update_timer()
    {
        timer_ = this->create_wall_timer(
            std::chrono::duration<double>(1.0 / params_.frequency), std::bind(&SineWavePublisher::publish_sine_wave, this));
    }

    rclcpp::Publisher<std_msgs::msg::Float64>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    double angle_;
    std::shared_ptr<sine_wave_params::ParamListener> param_listener_;
    sine_wave_params::Params params_;
    rclcpp::node_interfaces::OnSetParametersCallbackHandle::SharedPtr param_callback_handle_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SineWavePublisher>());
    rclcpp::shutdown();
    return 0;
}